test/Microsoft.AspNetCore.Blazor.Build.Test/TestFiles/RuntimeCodeGenerationTest/ChildComponent_GenericWeaklyTypedAttribute_TypeInference/TestComponent.codegen.cs

```diff
@@ -0,0 +1,37 @@
+// <auto-generated/>
+#pragma warning disable 1591
+namespace Test
+{
+    #line hidden
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore.Blazor;
+    using Microsoft.AspNetCore.Blazor.Components;
+    public class TestComponent : Microsoft.AspNetCore.Blazor.Components.BlazorComponent
+    {
+        #pragma warning disable 1998
+        protected override void BuildRenderTree(Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeBuilder builder)
+        {
+            base.BuildRenderTree(builder);
+            __Blazor.Test.TestComponent.TypeInference.CreateMyComponent_0(builder, 0, 1, "hi", 2, 17);
+        }
+        #pragma warning restore 1998
+    }
+}
+namespace __Blazor.Test.TestComponent
+{
+    #line hidden
+    internal static class TypeInference
+    {
+        public static void CreateMyComponent_0<TItem>(global::Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeBuilder builder, int seq, int __seq0, TItem __arg0, int __seq1, System.Object __arg1)
```

Comment:
Is it possible that this "treat weakly-typed args as `System.Object`" breaks our use of different overloads of `AddAttribute` for different arg types?

Currently we have overloads of `AddAttribute` for `bool`, `string`, `Action`, `Action<UIEventArgs>`, `Func<Task>`, `Func<UIEventArgs, Task>`, `MulticastDelegate`, `object`, and `RenderTreeFrame`. Each has slightly different logic for converting the arg, skipping if it's a false boolean, etc.

Reply:
Actually on reading the `AddAttribute` codes more closely, it's not necessarily problematic, since the considerable majority of the special cases are skipped in the case of components (rather than elements).

However structuring the code like this, where each `AddAttribute` overload separately makes component-vs-element decisions seems like it's inviting misunderstanding/bugs. I wonder if we should add a new `AddComponentParameter` to `RenderTreeBuilder`, whose logic is much simpler, and then we can call that from here and anywhere else that we know we're in the context of a component. Then we can change the original `AddAttribute` just to throw if it gets used for a component (maybe even rename all its overloads to `AddElementAttribute`).

If you like that and want to do it, that's great. If you like that and don't want to do it now, fine too, maybe we could have a backlog item to remember it. Or if you just don't like it, let me know!

Reply:
It would also be good to split into `AddComponentParameter`/`AddElementAttribute` just so we're not doing so many `if` tests and branching at runtime, given we could have resolved all that at compile-time.

Reply:
So weak typing can absolutely break type inference/or be broken when using type inference. This is a limitation of how the compiler works.  For delegates/bind/eventhandlers I'm actually using `Action<object>` which resolves a tricky case. 

I'm going to log a backlog item to consider changing the runtime in the future. I'm not sure how it helps this feature without know that specific things are broken..